src/main/java/com/google/collegeplanner/servlets/PlannerServlet.java

```diff
@@ -29,35 +35,209 @@
+            }
+          }
+        }
+        // Breaks out if there are no more required classes and we exceeded our credit limit
+        if (depths.get(course) != i && avgCredits <= 0) {
+          break;
+        }
+      }
+      // Update indegrees for all next courses
+      for (String course : semester) {
+        for (String nextCourse : nextCourses.get(course)) {
+          indegree.replace(nextCourse, indegree.get(nextCourse) - 1);
+        }
+      }
+      semesters.add(semester);
+    }
+    // If courseList is not empty, scheduling was not possible
+    if (!courseList.isEmpty()) {
+      respondWithError(
+          "Not enough semesters for schedule.", HttpServletResponse.SC_BAD_REQUEST, response);
+    }
     response.setContentType("applications/json;");
-    response.getWriter().println(new Gson().toJson(courses));
+    response.getWriter().println(new Gson().toJson(semesters));
+  }
+
+  /**
+   * Given a course, returns whether all of its corequisites have a indegree of 0
+   *
+   * @param course The course ID for the course
+   */
+  private boolean isValidToAdd(String course) {
+    for (String coreq : corequisites.get(course)) {
+      if (indegree.get(coreq) != 0) {
+        return false;
+      }
+    }
+    return true;
   }
 
-  private static JSONObject getBody(HttpServletRequest request)
+  /**
+   * Gets the JSON Representation of the body of the POST request
+   */
+  private JSONObject getBody(HttpServletRequest request)
       throws IOException, ParseException, NullPointerException {
     String strBody =
         request.getReader().lines().collect(Collectors.joining(System.lineSeparator()));
     JSONParser parser = new JSONParser();
     return (JSONObject) parser.parse(strBody);
   }
 
+  /**
+   * Creates the representation for the graph and reads/stores necessary information from
+   * selectedClasses
+   *
+   * @param selectedClasses JSONArray that contains details for all of the courses selected by the
+   *     user
+   */
+  private void constructGraphs(JSONArray selectedClasses)
+      throws IOException, NumberFormatException {
+    // Initialize graph representation and courseList
+    indegree = new HashMap<String, Integer>();
+    corequisites = new HashMap<String, HashSet<String>>();
+    depths = new HashMap<String, Integer>();
+    credits = new HashMap<String, Integer>();
+    nextCourses = new HashMap<String, HashSet<String>>();
+    courseList = new ArrayList<>();
+
+    // creates CourseList and initialize nextCourses HashSets
+    for (Object course : selectedClasses) {
+      String key = (String) ((JSONObject) course).get("course_id");
+      courseList.add(key);
+      nextCourses.put(key, new HashSet<>());
+    }
+
+    // populate class variables
+    for (Object course : selectedClasses) {
+      String key = (String) ((JSONObject) course).get("course_id");
+      int creditVal = Integer.parseInt((String) ((JSONObject) course).get("credits"));
```

Comment:
It seems like you're getting the full data object from the frontend. Note that this is not really best practice. Consider this example:

1. User selects courses.
2. Database updates with new prerequisite data.
3. User generates a schedule.

The user will wind up using out-of-date data. Instead, we should only be passing the course_id as an input, and we should do a database lookup.

Of course, we'll need to defer this until we get datastore working though, but it would be good to file an issue to handle this.